htlcswitch/iterator.go

```diff
@@ -194,32 +239,141 @@ func (p *OnionProcessor) DecodeHopIterator(r io.Reader, rHash []byte) (HopIterat
 		}
 	}
 
-	return &sphinxHopIterator{
-		nextPacket:      sphinxPacket.NextPacket,
-		processedPacket: sphinxPacket,
-	}, lnwire.CodeNone
+	return makeSphinxHopIterator(onionPkt, sphinxPacket), lnwire.CodeNone
 }
 
-// ExtractErrorEncrypter takes an io.Reader which should contain the onion
-// packet as original received by a forwarding node and creates an
-// ErrorEncrypter instance using the derived shared secret. In the case that en
-// error occurs, a lnwire failure code detailing the parsing failure will be
-// returned.
-func (p *OnionProcessor) ExtractErrorEncrypter(r io.Reader) (ErrorEncrypter, lnwire.FailCode) {
+// DecodeHopIterators performs batched decoding and validation of incoming
+// sphinx packets. For the same `id`, this method will return the same iterators
+// and failcodes upon subsequent invocations.
+//
+// NOTE: In order for the responses to be valid, the caller must guarantee that
+// the presented readers and rhashes *NEVER* deviate across invocations for the
+// same id.
+func (p *OnionProcessor) DecodeHopIterators(id []byte, rs []io.Reader,
+	rHashes [][]byte, incomingCltvs []uint32) ([]HopIterator,
+	[]lnwire.FailCode, error) {
 
-	onionPkt := &sphinx.OnionPacket{}
-	if err := onionPkt.Decode(r); err != nil {
+	batchSize := len(rs)
+
+	var (
+		onionPkts = make([]sphinx.OnionPacket, batchSize)
```

Comment:
It seems that all of these should be something like `make([]T, 0, batchSize)`, as it's possible that some of the iterators or w/e don't get created due to a partial failure. Otherwise, seems like the caller would keep iterating and potentially try to parse an object that hasn't been fully populated.

Reply:
I have a comment about it in lightning-onion [here](https://github.com/lightningnetwork/lightning-onion/blob/master/sphinx.go#L825). Admittedly, I remember it being in a more prominent place 😂I think a similar comment should be added to the top of this function, specifying an entry is only valid if its fail code is `lnwire.CodeNone`.

FWIW, the primary reason we do this is because the granularity of packets that get processed here always corresponds to the Adds in a particular forwarding package. If we filter out replayed packets, we would need an additional layer of translation between the indexes being tracked by forwarding packages and sphinx replay sets. The sparse list allows us to eliminate the need for doing this translation, and ensures that our indexing is consistent across the subsystems.

Reply:
If a packet is failed during a replay, we still have actions to perform at the link-level in order to fail it back. So the user should still be iterating over the full list to check for these failures. The goal was to provide batched version of a double-value context, e.g. `val, err := somethingThatCanFail()`, where `val` should only be accessed if `err` is not nil.